### PR TITLE
add default slot and keys recognition

### PR DESCRIPTION
**Pull request by anatolysergeev**
add default connection for commands without a key and for multiple key commands with the limitation that all keys have to be in the same slot

#673

### CODE REVIEW

File: redis/src/main/scala/zio/redis/Input.scala [hunk 1]
```diff
@@ -526,114 +573,114 @@ object Input {
...
   }
 
   final case class Varargs[-A](input: Input[A]) extends Input[Iterable[A]] {
-    def encode(data: Iterable[A])(implicit codec: BinaryCodec): Chunk[RespValue.BulkString] =
-      data.foldLeft(Chunk.empty: Chunk[RespValue.BulkString])((acc, a) => acc ++ input.encode(a))
+    def encode(data: Iterable[A])(implicit codec: BinaryCodec): RespCommand =
+      data.foldLeft(RespCommand.empty)((acc, a) => acc ++ input.encode(a))
   }
 
   final case class EvalInput[-K, -V](inputK: Input[K], inputV: Input[V]) extends Input[(String, Chunk[K], Chunk[V])] {
```
Comment (anatolysergeev) [inline redis/src/main/scala/zio/redis/Input.scala:632]: Controversial idea. What Is interface ideally wanted: inputK would have type KeyInput[K] that inherit Input[K]  and we can translate there any dependency, but we cannot to do that cause we have to have contramap that in the inherited interface that return main trait Input. 🤔 
@mijicd What do you think?

File: redis/src/main/scala/zio/redis/ClusterExecutor.scala [hunk 2]
```diff
@@ -58,8 +58,8 @@ final case class ClusterExecutor(
     }
 
     for {
-      key    <- ZIO.attempt(command(1)).orElseFail(CusterKeyError)
-      keySlot = Slot((key.asCRC16 % SlotsAmount).toLong)
+      keyOpt <- ZIO.succeed(command.respArgs.collectFirst { case key: RespArgument.Key => key })
+      keySlot = keyOpt.fold(Slot.Default)(key => Slot((key.asCRC16 % SlotsAmount).toLong))
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/ClusterExecutor.scala:62]: Since `SlotsAmount` is 2^14, you can replace `%` with `&`, i.e. `key.asCRC16 & (SlotsAmount - 1)`. The same applies for other usages of `%` with `SlotsAmount` (if any).

File: redis/src/main/scala/zio/redis/RespValue.scala [hunk 3]
```diff
@@ -17,12 +17,78 @@
...
+  def apply(respArgs: Chunk[RespArgument]): RespCommand = new RespCommand(respArgs)
+
+  def apply(respArgs: RespArgument*): RespCommand = new RespCommand(Chunk.fromIterable(respArgs))
+
+  def apply(respArg: RespArgument): RespCommand = new RespCommand(Chunk.single(respArg))
+}
+
+sealed trait RespArgument {
+  def respValue: RespValue.BulkString
+}
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/RespValue.scala:48]: If you would like to proceed with this, please:
- submit it in a separate pull request as it is not related to slots and keys
- either split it to multiple files or respect scala's file-naming convention
- benchmark, as it will introduce some penalty

File: redis/src/main/scala/zio/redis/RespValue.scala [hunk 4]
```diff
@@ -17,12 +17,78 @@
...
+
+object RespArgument {
+
+  case class Unknown(bytes: Chunk[Byte]) extends RespArgument {
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/RespValue.scala:52]: Case classes should be final.

**Review by mijicd — CHANGES_REQUESTED** [review]
Please split the pull request as suggested, and I'll re-review :pray:

File: redis/src/main/scala/zio/redis/RespValue.scala (hunk 3 above)
Comment (anatolysergeev) [reply redis/src/main/scala/zio/redis/RespValue.scala:48]: I did few benchmarks:
`[info] Benchmark                    (count)   Mode  Cnt   Score   Error  Units
[info] AppendBenchmarks.zio             500  thrpt   30   9.986 ± 0.026  ops/s ## my branch
[info] AppendBenchmarks.zio             500  thrpt   30  10.071 ± 0.035  ops/s

[info] Benchmark          (count)   Mode  Cnt   Score   Error  Units
[info] SetBenchmarks.zio      500  thrpt   30   9.918 ± 0.025  ops/s           ## my branch
[info] SetBenchmarks.zio      500  thrpt   30  10.426 ± 0.037  ops/s

[info] Benchmark          (count)   Mode  Cnt  Score   Error  Units
[info] GetBenchmarks.zio      500  thrpt   30  9.358 ± 0.631  ops/s            ## my branch
[info] GetBenchmarks.zio      500  thrpt   30  9.927 ± 0.023  ops/s

[info] Benchmark           (count)   Mode  Cnt   Score   Error  Units
[info] IncrBenchmarks.zio      500  thrpt   30   9.931 ± 0.203  ops/s          ## my branch
[info] IncrBenchmarks.zio      500  thrpt   30  10.188 ± 0.203  ops/s`

Is it enough?

Comment (anatolysergeev) [reply redis/src/main/scala/zio/redis/RespValue.scala:48]: >submit it in a separate pull request as it is not related to slots and keys

It's related to slot and keys, we have to differentiate parameter types in executor

Comment (anatolysergeev) [reply redis/src/main/scala/zio/redis/RespValue.scala]: >either split it to multiple files or respect scala's file-naming convention

Splitted

Comment (mijicd) [reply redis/src/main/scala/zio/redis/RespValue.scala]: > >submit it in a separate pull request as it is not related to slots and keys
> 
> It's related to slot and keys, we have to differentiate parameter types in executor

It causes a ripple effect of changes throughout the entire library. I would argue that this is not a "local" change :)

Comment (anatolysergeev) [reply redis/src/main/scala/zio/redis/RespValue.scala]: I agree with you it is not a "local" change - there is no argues.) I don't know how to solve that issue do not change inner architecture 😇 
But current architecture do not have tools to differentiate Redis (resp) parameter types, that is necessary for advanced client library. If we want to have flexible, easy maintainable executors we need it. Something similar is implemented in the lettuce, i get basic idea from it.

Comment (mijicd) [reply redis/src/main/scala/zio/redis/RespValue.scala]: I am with you when it comes to the approach to solve it, just do the RespCommand part in one PR and "apply" it another :)

Comment (anatolysergeev) [reply redis/src/main/scala/zio/redis/RespValue.scala]: Do you want that?) 
https://github.com/zio/zio-redis/pull/763

File: redis/src/test/scala/zio/redis/RespArgumentSpec.scala [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+import zio.test._
+
+object RespArgumentSpec extends BaseSpec {
+  def spec: Spec[Any, RedisError.ProtocolError] =
+    suite("RespArgument")(
+      suite("BulkString.asCRC16")(
+        test("key without braces") {
+          val key = RespArgument.Key("hello world")
+          assertTrue(15332 == key.asCRC16)
```
Comment (mijicd) [inline redis/src/test/scala/zio/redis/RespArgumentSpec.scala:11]: Nitpick (I'm fine with leaving it as it is as well):

```suggestion
          assertTrue(15332 == RespArgument.Key("hello world").asCRC16)
```

**Review by mijicd — APPROVED** [review]
